crates/pera-core/src/signature_mpc/mpc_manager.rs

```diff
@@ -0,0 +1,450 @@
+    ) -> PeraResult<()> {
+        let Some(mut instance) = self.mpc_instances.get_mut(&session_id) else {
+            error!(
+                "received a message for instance {:?} which does not exist",
+                session_id
+            );
+            // TODO (#261): Punish a validator that sends a message related to a non-existing mpc instance
+            return Ok(());
+        };
+        instance.handle_message(SignatureMPCMessage {
+            message: message.to_vec(),
+            authority: authority_name,
+        })
+    }
+
+    /// Spawns a new MPC instance if the number of active instances is below the limit
+    /// Otherwise, adds the instance to the pending queue
+    fn push_new_mpc_instance(&mut self, event: P::InitEvent) {
+        if self.mpc_instances.contains_key(&event.session_id().bytes) {
+            // This should never happen, as the session ID is a move UniqueID
+            error!(
+                "Received start flow event for session ID {:?} that already exists",
+                event.session_id()
+            );
+            return;
+        }
+
+        info!(
+            "Received start flow event for session ID {:?}",
+            event.session_id()
+        );
+
+        if self.active_instances_counter > self.max_active_mpc_instances {
+            self.pending_instances_queue
+                .push_back(event.session_id().bytes);
+            info!(
+                "Added MPCInstance to pending queue for session_id {:?}",
+                event.session_id()
+            );
+            return;
+        }
+        let new_instance = SignatureMPCInstance::new(
+            Arc::clone(&self.consensus_adapter),
+            self.epoch_store.clone(),
+            self.epoch_id,
+            event.session_id().clone().bytes,
+            event.event_emitter().clone(),
+            self.number_of_parties,
+        );
+        self.mpc_instances
+            .insert(event.session_id().clone().bytes, new_instance);
+        self.active_instances_counter += 1;
+        info!(
+            "Added MPCInstance to MPC manager for session_id {:?}",
+            event.session_id()
+        );
+    }
+
+    fn finalize_mpc_instance(&mut self, event: P::FinalizeEvent) -> PeraResult {
+        let session_id = event.session_id().bytes;
+        let mut instance = self.mpc_instances.get_mut(&session_id).ok_or_else(|| {
+            PeraError::InvalidCommittee(format!(
+                "Received a finalize event for session ID {:?} that does not exist",
+                event.session_id()
+            ))
+        })?;
+        if let MPCSessionStatus::Finalizing(output) = &instance.status {
+            instance.status = MPCSessionStatus::Finished(output.clone());
+            self.active_instances_counter -= 1;
+            info!(
+                "Finalized MPCInstance for session_id {:?}",
+                event.session_id()
+            );
+            return Ok(());
+        }
```

Comment:
Should this not pop the next pending instance from the queue?

Reply:
Will be handled here:
https://github.com/dwallet-labs/dwallet-network/issues/282

Reply:
I don't know if this is the best place to put it, we will think about it.
It's part of this issue:
https://github.com/dwallet-labs/dwallet-network/issues/307